### PR TITLE
DAOS-1939 mgmt: implement handling of drpc daos requests

### PR DESCRIPTION
**Pull request by tanabarr**
Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/mgmt/srv.c [hunk 1]
```diff
@@ -59,23 +59,71 @@ static struct daos_rpc_handler mgmt_handlers[] = {
...
+
+			proto__daos_rank__free_unpacked(pb_rank, NULL);
+		}
+	} else {
+		drpc_resp->status = DRPC__STATUS__UNKNOWN_METHOD;
+		D_ERROR("Unknown method\n");
+	}
+
+	/* Populate drpc response body with daos response. TODO: verify len */
+	size_t len = proto__daos_response__pack(daos_resp, body);
```
Comment (kjacque) [inline src/mgmt/srv.c]: Don't want to allocate the body in all cases, right? If there was a bad method requested, for example.

File: src/mgmt/srv.c [hunk 2]
```diff
@@ -59,23 +59,71 @@ static struct daos_rpc_handler mgmt_handlers[] = {
...
+	drpc_resp->status = DRPC__STATUS__SUCCESS;
+
+	/* TODO: select against multiple methods */
+	if (request->method == DRPC_METHOD_MGMT_KILL_RANK) {
+		/* Unpack the daos request from the drpc call body */
+		pb_rank = proto__daos_rank__unpack(
+			NULL, request->body.len, request->body.data);
+
+		if (pb_rank == NULL) {
+			drpc_resp->status = DRPC__STATUS__FAILURE;
```
Comment (kjacque) [inline src/mgmt/srv.c]: I'm wondering if this failure should set a method-specific status instead. Feels like it sort of skirts the line. It is an invalid drpc message for this method, but the expectations really belong to the method. Plus the DaosResponse could offer a more detailed status. I could be swayed the other way, though.

Comment (tanabarr) [reply src/mgmt/srv.c]: yes I was thinking about this, ideally we want a generic DaosRequest which is unpacked and passed to the daos request handler, if the DaosRequest is not found in the drpc request body than that constitutes as a drpc failure and we should respond with failure status and empty DaosResponse in body of drpc response. what I have done currently is delegate the knowledge of the request format to the method specific handler code and if the request fails to unpack constitutes as drpc failure. current solution not very tidy. I get what you are saying but don't think we should be including method specific errors in drpc response.

File: src/mgmt/srv.c (hunk 1 above)
Comment (tanabarr) [reply src/mgmt/srv.c]: I was always allocating body even in the case of failure, can change that though

File: src/mgmt/srv.pb-c.c [hunk 3]
```diff
@@ -0,0 +1,1434 @@
...
+#endif
+
+#include "srv.pb-c.h"
+void   proto__empty_params__init
+                     (Proto__EmptyParams         *message)
+{
+  static const Proto__EmptyParams init_value = PROTO__EMPTY_PARAMS__INIT;
+  *message = init_value;
+}
+size_t proto__empty_params__get_packed_size
```
Comment (dpquigl) [inline src/mgmt/srv.pb-c.c]: Im not sure why these are prefaced with proto while the security ones aren't when they both have proto as the package. I think it has to do with the path used to generate the C source fils. I'll have to look into it.

File: src/mgmt/srv.c (hunk 2 above)
Comment (kjacque) [reply src/mgmt/srv.c]: Definitely agreed on the last point. Drpc__Response status should be used to indicate higher-level failures that occur before really running the requested method, and any method-specific failure codes should be in the DaosResponse marshaled into the response body.

Thinking more about it, a corrupted body on the incoming message (which would prevent unmarshaling) does seem like a protocol problem, and it prevents us from doing anything even remotely useful in the method. So Drpc__Response is probably the right place to indicate that failure.

Comment (tanabarr) [reply src/mgmt/srv.c]: glad you agree with the implementation 👍

[the remark this answers is not in the document]
Comment (tanabarr) [reply src/mgmt/srv.c]: as discussed below, unexpected payload formats indicate a protocol error if the payload cannot be passed to the next layer. therefore if the request cannot be decoded it makes sense that drpc reports the failure IMO. if the problem occurs during processing of the embedded request (i.e. the daos command) then that failure should be represented in the status of the embedded response

File: Jenkinsfile [hunk 4]
```diff
@@ -85,7 +85,7 @@ pipeline {
                     steps {
                         checkPatch user: GITHUB_USER_USR,
                                    password: GITHUB_USER_PSW,
-                                   ignored_files: "src/control/vendor/*"
+                                   ignored_files: "src/control/vendor/*,*.pb.[ch]"
```
Comment (brianjmurrell) [inline Jenkinsfile]: It seems like `,` would be a good separator and maybe should be accepted, but currently `:` is the only accepted separator.

```suggestion
                                   ignored_files: "src/control/vendor/*:*.pb.[ch]"
```

**Comment by tanabarr** [comment]
moved to #259 to get rid of checkpatch warnings on generated protobuf files

**Pull request closed without merging**